### ISSUE TITLE
werkzeug.test.stream_encode_multipart raises ValueError if the data contains empty stream

### ISSUE THREAD

**Issue by pandabear** [pallets/werkzeug#2740]
This was discovered in tests pipeline upon upgrading to latest Flask and during pytest run. `FlaskClient` was used to upload an empty bytes file with something like
```
resp = test_client.post(
    '/my/route',
    data={'file': (b'', 'test.json')}
)
```
It will trigger a ValueError with the following stack trace:
```
venv39/lib64/python3.9/site-packages/flask/testing.py:184: in _request_from_builder_args
    return builder.get_request()
venv39/lib64/python3.9/site-packages/werkzeug/test.py:795: in get_request
    return cls(self.get_environ())
venv39/lib64/python3.9/site-packages/werkzeug/test.py:719: in get_environ
    input_stream, content_length, boundary = stream_encode_multipart(
venv39/lib64/python3.9/site-packages/werkzeug/test.py:131: in stream_encode_multipart
    encoder.send_event(

self = <werkzeug.sansio.multipart.MultipartEncoder object at 0x7f23c813b220>, event = File(name='info', filename='test.json', headers=Headers([('Content-Type', 'application/json')]))

    def send_event(self, event: Event) -> bytes:
        if isinstance(event, Preamble) and self.state == State.PREAMBLE:
            self.state = State.PART
            return event.data
        elif isinstance(event, (Field, File)) and self.state in {
            State.PREAMBLE,
            State.PART,
            State.DATA,
        }:
            data = b"\r\n--" + self.boundary + b"\r\n"
            data += b'Content-Disposition: form-data; name="%s"' % event.name.encode()
            if isinstance(event, File):
                data += b'; filename="%s"' % event.filename.encode()
            data += b"\r\n"
            for name, value in t.cast(Field, event).headers:
                if name.lower() != "content-disposition":
                    data += f"{name}: {value}\r\n".encode()
            self.state = State.DATA_START
            return data
        elif isinstance(event, Data) and self.state == State.DATA_START:
            self.state = State.DATA
            if len(event.data) > 0:
                return b"\r\n" + event.data
            else:
                return event.data
        elif isinstance(event, Data) and self.state == State.DATA:
            return event.data
        elif isinstance(event, Epilogue):
            self.state = State.COMPLETE
            return b"\r\n--" + self.boundary + b"--\r\n" + event.data
        else:
>           raise ValueError(f"Cannot generate {event} in state: {self.state}")
E           ValueError: Cannot generate File(name='info', filename='test.json', headers=Headers([('Content-Type', 'application/json')])) in state: State.DATA_START

venv39/lib64/python3.9/site-packages/werkzeug/sansio/multipart.py:313: ValueError
```

I would've expected that an empty file could've been uploaded as multipart just fine.

Upon further examination, the cause seem to be this line: https://github.com/pallets/werkzeug/blob/82b5ac433c54b8d03cb0de9fdc16a4c58b935860/src/werkzeug/test.py#L135-L141 . The `encoder`'s state never gets updated from `State.DATA_START`, since the `chunk` always returns `Falsey` and thus never gets to conduct the necessary `encoder.send_event`-call.

So importing `from .sansio.multipart import State` and the modifying the line `if not chunk: ` -> `if not chunk and encoder.state != State.DATA_START:` seemed to have solved this.

Environment:

- Python version: 3.9
- Werkzeug version: 2.3.6

**Comment by davidism** [comment]
Happy to review a PR

**Comment by pandabear** [comment]
PR created here: https://github.com/pallets/werkzeug/pull/2742 !

**Comment by pgjones** [comment]
Closed by #2765.
